src/chrome/content/ff-overlay.js

```diff
@@ -85,30 +85,37 @@ if (typeof SL_Overlay === "undefined") {
                 .addEventListener("popupshowing", (event) => {
                     this.showFirefoxContextMenu(event);
                 }, false);
-            DownloadManager.connectToNas();
         };
 
         this.onMenuItemLinkCommand = (event) => {
-            if (DownloadManager.isConnected) {
-                window.open(DownloadManager.urlToConnect + "/webman/index.cgi?launchApp=SYNO.SDS.DownloadStation.Application", "Diskstation", this.strWindowFeatures);
-            } else {
-                Notification.show("No Connection");
-            }
+            DownloadManager.connectAndRun(() => {
+                if (DownloadManager.isConnected) {
+                    window.open(DownloadManager.urlToConnect + "/webman/index.cgi?launchApp=SYNO.SDS.DownloadStation.Application", "Diskstation", this.strWindowFeatures);
+                } else {
+                    Notification.show("No Connection");
+                }
+            });
         };
 
         this.onMenuItemCommand = (event) => {
-            if (DownloadManager.isConnected) {
-                DownloadManager.transferToNas(gContextMenu.linkURL);
-            } else {
-                Notification.show("No Connection");
-            }
+            let linkURL = gContextMenu.linkURL;
+            DownloadManager.connectAndRun(() => {
```

Comment:
@Lemutar, I've tried to move it all into DownloadManager but have come across the same problems that I initially ran into, leading me to do it the way I did.
The problem is that the function that gets called (i.e. `transferToNas`, `updateListPanel`, etc.), needs to be passed to `connectToNas` and called after setting up a successful connection. If the connection is already set up though, the function can simply be called.

I don't see how to do this elegantly without a helper function (like `connectAndRun`) that handles this nicely.

Would appreciate your input on this one, I just can't see it at the moment :grimacing: 
Thanks!